Add method inlining to inline constructors filter

basro opened the pull request:
Normally inline constructors would immediately cancel if a field that is not a member variable is accessed.

For example:
```haxe
class MyIterator {
	public var i : Int = 0;
	public inline function new() {}
	public inline function hasNext() return true;
	public inline function next() return i++;
}

// ...

var a = new MyIterator();
a.hasNext(); // The call to hasNext is inlined before the inline constructors runs, no problem here.

var b : Iterator<Int> = a;
b.hasNext(); // Oops, hasNext was not inlined beforehand. 'hasNext' is not a member variable of MyIterator so inlining is cancelled.
```

With this PR the inline constructor filter will not give up so easily.

When the filter finds an expression of the form `ethis.method(args)` if `ethis` references an inline object and `method` is an inline method in the inline object then it will be inlined on demand by the filter.

This achieves the goals mentioned in #9432

It also probably solves #3147 (requires implementing proper inlineable iterator for map types and fixing the TFor issue I discussed with simn in slack)

I tested using the IntMap iterator from #8806 and this code:
```haxe
var m = [1 => 2, 3 => 4];
var iter = m.iterator();
while(iter.hasNext()) {
	var i = iter.next();
	trace(i);
}
```
Compiled into js:
```javascript
var iter_map_h = { };
iter_map_h[1] = 2;
iter_map_h[3] = 4;
var a = [];
for( var key in iter_map_h ) (iter_map_h.hasOwnProperty(key) ? a.push(key | 0) : null);
var keys = a;
var iter_keys = keys;
var iter_index = 0;
var iter_count = keys.length;
while(iter_index < iter_count) {
	var i = iter_map_h[iter_keys[iter_index++]];
	console.log("src/Main.hx:31:",i);
}
```

## Explanations about some of the changes

### Identifying inline objects

Previously inline objects were being identified by the order in which they appeared in the code. While it sounds nice in theory, nested inline objects inside inlined constructors make this very problematic. It requires keeping track of the start and end id of every inlined constructor to be able to reproduce the same id's in the final_pass phase.

In this PR the expressions are now mapped with a function called `mark_ctors` which will wrap any `TNew` `TArrayDecl` or `TObjectDecl` with @:inlineObject(curr_io_id++). This makes identifying inline objects in final_map a lot simpler, but it's a bit less efficient (Type.map_expr vs Type.iter).

### handle_field_case

In `analyse_aliases`, the function `handle_field_case` is used to handle analysis of TField and TArray. It used to return either None or Some inline variable that represents a field of an inline object.

`handle_field_case` now has a third possible return value IOFInlineMethod.

In the cases of TArray and TField this third value is ignored and causes cancellation of the inline object (Keeping the same behaviour it used to have).

IOFInlineMethod is only accepted without cancelling the inline object in the new `TCall({eexpr=TField(ethis,fa)},ca)` case of `analyze_aliases`.

basro committed:
Remove unused decls array

basro committed:
Mark inline objects using metadata instead of identifying them by order.

basro committed:
Remove unused e.etype from match

basro committed:
Replace e with mark_ctors result.

basro committed:
Fix mark_ctors

basro committed:
Remove unused id_start and id_end

basro committed:
debug hacks

basro committed:
Make handle field case handle inline methods too.

basro committed:
Implement inline constructor method inlining (Has known bugs marked with TODO)

basro committed:
Remove debug prints.

basro committed:
Fix typing of ethis on method inline.

basro committed:
For now don't forward captured to the inlined method analysis

basro committed:
Add missing result of this expression to inlined method.

basro committed:
Improve names of some variables.

basro committed:
Analyze aliases in call args of inlined methods and cancel the resulting inline objects if the parent inline object is cancelled.

basro committed:
Remove commented code.

basro committed:
Support returning inline objects from inlined methods

basro committed:
remove debug stuff.

basro committed:
Factor out call arg analysis

basro committed:
Some documentation.

basro committed:
Add return type to mark_ctors

basro committed:
Set io_has_untyped if inlined method had HasUntyped

basro committed:
Add @:inlineObject to meta.json and use that instead of Meta.Custom

basro committed:
Add more comments

basro committed:
Add inline constructor tests.

basro commented:
I've added a few tests.

Simn commented on src/optimization/inlineConstructors.ml:
```diff
@@ -42,47 +41,69 @@ open Globals
+		tclass *
+		Type.tparams *
+		tclass_field *
+		bool (* Wether the inline constructor is forced or not. Cancelling a forced constructor should produce an error *)
```
I recommend using a record type instead of a tuple type here, in my experience that's easier to work with and also easier to modify.

basro replied:
I've committed this change.

RealyUniqueName commented on src/optimization/inlineConstructors.ml:
```diff
@@ -329,41 +423,67 @@ let inline_constructors ctx e =
+			handle_default_case e
 	in
+	let original_e = e in
+	let e = mark_ctors e in
```
Most methods

RealyUniqueName commented on src/optimization/inlineConstructors.ml:
```diff
@@ -42,47 +41,69 @@ open Globals
+	| IOFInlineVar of inline_var
+	| IOFNone
+
 let inline_constructors ctx e =
```
Most methods in Haxe code probably don't contain inlinable objects/arrays. 
Wouldn't it be more efficient to run `check_expr` on `e` before doing any actual work to find out if the expression has something to inline.

basro replied:
This is a good idea.

Instead of `check_expr`, wouldn't it be better to check if `curr_io_id` is not 0?
In contrast to `check_expr` in the case that no inline objects are found it will have the cost of a noop Type.map_expr. But when objects are found it saves the cost of an extra iteration through the expression tree.

basro replied:
On second thought, using `check_expr` is probably better since it can short circuit and ends as soon as one inline object is found.

RealyUniqueName replied:
Yes, that's what I'm talking about.

basro replied:
I've committed this change.

RealyUniqueName commented on tests/optimization/src/TestInlineConstructors.hx:
```diff
@@ -0,0 +1,92 @@
+class InlineClass {
```
Missing a test with an assignment to anon structure (provided in PR message).

basro replied:
I don't understand what you mean, doesn't `testAnonymousStructureInlining` test for that?

RealyUniqueName replied:
No, I mean
```haxe
var a = new Inlinable();
a.job();
var b:{function job():Void;} = a;
b.job();
```

basro replied:
What about `testMethodInlining`?

RealyUniqueName replied:
Oh, there it is :)

basro committed:
Use record for IOKCtor data

basro committed:
Check that the expression has inline objects before doing any work.

basro committed:
Make sure the field we are inlining is a compatible function

This can only happen when dynamic casts are involved.
This is probably over-conservative, instead of type_eq_strict it would be better to check if it unifies.

basro committed:
Add test for incompatible method inline cancellations.

basro commented:
I found a bug (fix is committed with tests)

The bug should only happen in weird dynamic casts:
```haxe
class InlineClass {
  public var a : Int;
  public inline function new() {};
  public inline function method() return a;
}

// ...

var a : {function method(arg:String) : Int; } = cast new InlineClass();
// InlineClass.method doesn't take any arguments.
// This should definitely cancel the inlining.
a.method("hello"); 
```

My fix uses type_iseq_strict to compare `a.method` type and the `InlineClass` method `cf_type`. If false then inlining is cancelled.

This is probably overly conservative, I think that checking if it unifies and that the efield type is not Dynamic would be better. I'm going to try that next.

basro committed:
Check that the class method type unifies with the field expression type instead.

Also check that the field expression is a function to avoid the Dynamic case.

RealyUniqueName commented:
Please update to the latest development

basro commented:
I've rebased on latest development
